d2/src/main/java/com/linkedin/d2/balancer/LoadBalancerState.java

```diff
@@ -73,6 +73,16 @@
 
   void listenToCluster(String clusterName, LoadBalancerStateListenerCallback callback);
 
+  /**
+   * Stops listening to a cluster
+   * @param clusterName the cluster to stop listening to.
+   * @return true if listeners successfully stopped.
+   */
+  default boolean stopListenToCluster(String clusterName)
```

Comment:
The signature of `listenToCluster` is taking a callback. I think the reason is that we can supply a `FutureCallback` and provide a timeout value in case this operation is blocking for too long: https://github.com/linkedin/rest.li/blob/ac4ad1f91b91c0f7f974d6af1f4b60b3bbcace2e/d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancer.java#L503.
Shall we do the same here by passing in a callback, and invoke `callback.done` when the subscriber is unregistered?

Reply:
In this case, do we expect caller to verify if the listener is stopped successfully? Or should I create another callback that returns the success/failure as part of the `callback.done`?

Reply:
I think the current logic looks good. But just that this `ensureNotListening` is not really ensuring (e.g. in case the queue is not closed). Maybe change the name to `stopListening` would better describe what it does.

Reply:
Good point. Renamed to `tryStopListening` since it's not guaranteed.